lib/src/scanner/mod.rs

```diff
@@ -507,6 +499,26 @@ impl<'r> Scanner<'r> {
         Ok(())
     }
 
+    /// Updates the metadata for a module specified by its fully-qualified name.
+    ///
+    /// If the `meta` argument is `None`, the metadata for the module is removed.
+    ///
+    /// See [`Scanner::module_meta`] for the reasoning behind choosing `Arc<_>`
+    pub fn set_module_meta(
+        &mut self,
+        module_full_name: &str,
+        meta: Option<&Arc<[u8]>>,
```

Comment:
I believe it's possible to pass `Option<&[u8]>` here. It's will be harder due to lifetime issues, but that's actually the desired API. The scanner doesn't really need to retain the metadata after the scan has finished, therefore a reference to the metadata should be enough. However, in case lifetimes makes this hard to do, I think that using `Rc` is enough. I don't see a reason for using `Arc` as the scanner is not `Send` nor `Sync` anyways. You are supposed to use one scanner per thread, and I don't see why anyone would need to pass the same metadata to multiple scanners running in different threads.

We can use `Rc` for now and try to change it to `&[u8]` in a future commit.

Reply:
After thinking about this a little bit more, I believe I was wrong. Passing `&[u8]` to `set_module_meta` may be impossible without unsafe code. However, there is an alternative design that makes possible to get rid of those smart pointers `Arc` and `Rc` from the API. Suppose that we have a type like this:

```rust
pub struct ScanOptions<'a> {
}

impl<'a> ScanOptions<'a> {
    fn set_metadata(&mut self, module_name: &str, meta: &'a [u8]) {
    }
}
```

A new function would be added to `Scanner`:

```rust
    pub fn scan_with_options<'a>(
        &'a mut self,
        data: &'a [u8],
        options: ScanOptions<'a>,
    ) -> Result<ScanResults<'a, 'r>, ScanError> {
       ...
    }
```

Then you can use it as:

```rust

let mut scanner = Scanner::new(rules)
let mut options = ScanOptions::default();

options.set_metadata("pe", pe_metadata));
options.set_metadata("elf", elf_metadata);

scanner.scan_with_options(data, options)
```

I think this makes everything simpler because you don't need to store the modules metadata in `Scanner` nor in `ScannerContext`, and you don't have a lifetimes nightmare. It also opens the door for adding more scan options in the future by adding more methods to `ScanOptions`.

Reply:
mb for not checking scanner already is `!Send + !Sync`, should have been `Rc<_>`

rewrote it according to your suggestions anyway in [b699aa0](https://github.com/VirusTotal/yara-x/pull/183/commits/b699aa0d18894e047c23839d2423a85869fbad11)